jbpm-persistence/jbpm-persistence-jpa/src/main/java/org/jbpm/persistence/processinstance/JPAProcessInstanceManager.java

```diff
@@ -132,49 +135,53 @@ public ProcessInstance getProcessInstance(long id, boolean readOnly) {
-	        }
-	        processInstance.setProcess( process );
-        }
-        if ( processInstance.getKnowledgeRuntime() == null ) {
-            Long parentProcessInstanceId = (Long) ((ProcessInstanceImpl) processInstance).getMetaData().get("ParentProcessInstanceId");
-            if (parentProcessInstanceId != null) {
-                kruntime.getProcessInstance(parentProcessInstanceId);
+        try {
+        	// Make sure that the cmd scoped entity manager has started
+        	ProcessPersistenceContextManager ppcm 
+        	    = (ProcessPersistenceContextManager) this.kruntime.getEnvironment().get( EnvironmentName.PERSISTENCE_CONTEXT_MANAGER );
+        	ppcm.beginCommandScopedEntityManager();
+        	
+            ProcessPersistenceContext context = ppcm.getProcessPersistenceContext();
+            ProcessInstanceInfo processInstanceInfo = (ProcessInstanceInfo) context.findProcessInstanceInfo( id );
+            if ( processInstanceInfo == null ) {
+                return null;
             }
-            processInstance.setKnowledgeRuntime( kruntime );
-            
-            ((ProcessInstanceImpl) processInstance).reconnect();
-            if (readOnly) {
-                internalRemoveProcessInstance(processInstance);
+            processInstance = (org.jbpm.process.instance.ProcessInstance)
+            	processInstanceInfo.getProcessInstance(kruntime, this.kruntime.getEnvironment(), readOnly);
+            if (!readOnly) {
+                processInstanceInfo.updateLastReadDate();
+                TransactionManagerHelper.addToUpdatableSet(txm, processInstanceInfo);            
+            }
+            if (((ProcessInstanceImpl) processInstance).getProcessXml() == null) {
+    	        Process process = kruntime.getKieBase().getProcess( processInstance.getProcessId() );
+    	        if ( process == null ) {
+    	            throw new IllegalArgumentException( "Could not find process " + processInstance.getProcessId() );
+    	        }
+    	        processInstance.setProcess( process );
+            }
+            if ( processInstance.getKnowledgeRuntime() == null ) {
+                Long parentProcessInstanceId = (Long) ((ProcessInstanceImpl) processInstance).getMetaData().get("ParentProcessInstanceId");
+                if (parentProcessInstanceId != null) {
+                    kruntime.getProcessInstance(parentProcessInstanceId);
+                }
+                processInstance.setKnowledgeRuntime( kruntime );
+                
+                ((ProcessInstanceImpl) processInstance).reconnect();
+                if (readOnly) {
+                    internalRemoveProcessInstance(processInstance);
+                }
+            }
+            return processInstance;
+        } finally {
+            if (!readOnly && processInstance != null) {
+                EventManagerProvider.getInstance().get().update(getInstanceViewFor(processInstance));
```

Comment:
So this is for situations when something goes wrong? In such cases we want to push updates anyway?

Reply:
it should be added as long as it's not readonly and not null and to avoid adding same line in few places adding that in finally block made sense. It must be done once the process instance was reconnected.